atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultParallelTaskExecutor.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.asts;
+
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.refreshable.Refreshable;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public final class DefaultParallelTaskExecutor implements ParallelTaskExecutor {
+    private final ExecutorService cachedExecutorService;
+    private final Refreshable<Duration> semaphoreAcquireTimeout;
```

Comment:
for timeout: would be good to say what expectations we have of it (probably implicit, but we want this to be at least as long as the longest task we expect to submit to the executor)